### PR TITLE
Moved sector_identifier checks outside

### PR DESCRIPTION
**Pull request by tpazderka**
Refactoring of `oic.oic.provider.Provider.do_clinet_registration` in order to decrease McCabe score.

- [ ] Any changes relevant to users are recorded in the `CHANGELOG.md`.
- [ ] The documentation has been updated, if necessary.

---

### CODE REVIEW

File: src/oic/oic/provider.py [hunk 1]
```diff
@@ -1354,44 +1357,11 @@ def do_client_registration(self, request, client_id, ignore=None):
...
-                            "invalid_configuration_parameter",
-                            descr="redirect_uri missing from "
-                                  "sector_identifiers"
-                        )
-
-            _cinfo["si_redirects"] = si_redirects
-            _cinfo["sector_id"] = si_url
+                _cinfo["si_redirects"], _cinfo["sector_id"] = self._verify_sector_identifier(request)
+            except InvalidSectorIdentifier as err:
+                return error_response("invalid_configuration_parameter",
```
Comment (lwm) [inline src/oic/oic/provider.py]: Suppose we can make this all on one line? Could be nitpicking ;)

File: src/oic/oic/provider.py [hunk 2]
```diff
@@ -1433,12 +1403,9 @@ def do_client_registration(self, request, client_id, ignore=None):
...
-            try:
-                n_keys = len(self.keyjar[client_id])
-                msg = "found {} keys for client_id={}"
-                logger.debug(msg.format(n_keys, client_id))
-            except KeyError:
-                pass
+            n_keys = len(self.keyjar.get(client_id, []))
+            if n_keys:
+                logger.debug("Found %s keys for client_id=%s", n_keys, client_id)
         except Exception as err:
```
Comment (lwm) [inline src/oic/oic/provider.py]: Can we be more specific than `Exception` for the  `self.keyjar.load_keys(request, client_id)` failure when we catch?

File: src/oic/oic/provider.py [hunk 3]
```diff
@@ -1501,6 +1468,39 @@ def verify_redirect_uris(registration_request):
...
+
+        :param request: Provider registration request
+        :return: si_redirects, sector_id
+        :raises: InvalidSectorIdentifier
+        """
+        si_url = request["sector_identifier_uri"]
+        try:
+            res = self.server.http_request(si_url)
+        except ConnectionError as err:
+            logger.error("{}", err)
```
Comment (lwm) [inline src/oic/oic/provider.py]: Can just be `logger.error(err)`?

File: src/oic/oic/provider.py [hunk 4]
```diff
@@ -1501,6 +1468,39 @@ def verify_redirect_uris(registration_request):
...
+            res = None
+
+        if not res:
```
Comment (lwm) [inline src/oic/oic/provider.py]: This `raise` could be moved into the `except` block above?

File: src/oic/oic/provider.py [hunk 5]
```diff
@@ -1501,6 +1468,39 @@ def verify_redirect_uris(registration_request):
...
+
+        logger.debug("sector_identifier_uri => %s", sanitize(res.text))
+
+        try:
+            si_redirects = json.loads(res.tezt)
+        except ValueError:
+            raise InvalidSectorIdentifier("Error deserializing sector_identifier_uri content")
+
+        if "redirect_uris" in request:
+            logger.debug("redirect_uris: %s". request["redirect_uris"])
```
Comment (lwm) [inline src/oic/oic/provider.py]: What's that full stop doing!? Is this some new cool syntax that I didn't know about? :smile:

File: src/oic/oic/provider.py [hunk 6]
```diff
@@ -1501,6 +1468,39 @@ def verify_redirect_uris(registration_request):
...
+            for uri in request["redirect_uris"]:
```
Comment (lwm) [inline src/oic/oic/provider.py]: Might be just my preference but I like to skip nested fors (feel free to ignore):

``` python
if not all([uri in request["redirect_uris"] for uri in si_redirects]):
    raise InvalidSectorIdentifier("redirect_uri missing from sector_identifiers")
```

**Comment by lwm** [comment]
> I would like to add some test to be sure it is working properly

:beers: :rainbow: :wine_glass: :rainbow_flag: :beers:

File: src/oic/oic/provider.py (hunk 5 above)
Comment (tpazderka) [inline src/oic/oic/provider.py]: Typos... typos everywhere...

File: src/oic/oic/provider.py (hunk 2 above)
Comment (tpazderka) [inline src/oic/oic/provider.py]: We should, I will look into this.

File: src/oic/oic/provider.py (hunk 4 above)
Comment (tpazderka) [inline src/oic/oic/provider.py]: Probably not, if you look at the previous code, it was doing the same thing. This catches the case when the response doesn't raise an exception but returns `None`.

File: src/oic/oic/provider.py (hunk 6 above)
Comment (tpazderka) [inline src/oic/oic/provider.py]: I always find dict comprehension unreadable (personal thing) so I will gladly go with the ignore on this one :)

File: src/oic/oic/provider.py (hunk 2 above)
Comment (tpazderka) [inline src/oic/oic/provider.py]: Mental note to myself: `KeyIOError` and `UnknownKeyType` are clearly visible as raised exceptions from called methods. Other could be possibly hidden.

**Comment by tpazderka** [comment]
Tests for the method added. It is marked as internal since this is something that shouldn't be changed.

The score went down by `7` so some more work is needed...:)

**Comment by tpazderka** [comment]
@lwm, could you please have a look so we can merge this?

**Pull request merged**